lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -579,55 +575,72 @@ dotOperandMfmaToLinearLayout(DotOperandEncodingAttr dotMfmaLayout,
   // 4) warpsPerCTA[mDim] == 1: This guarantees that every B tensor element is
   //    held by exactly one thread, maintaining the same number of global loads
   //    as in a blocked layout.
+  //
+  // Other use of Linear layout is a support of rare corner cases,
+  // for example one instrcution tile is larger than tensor
   auto mfmaLayout = llvm::cast<AMDMfmaEncodingAttr>(dotMfmaLayout.getParent());
 
-  if (dotMfmaLayout.getOpIdx() == 0) {
-    return std::nullopt;
-  }
   auto rank = shape.size();
   bool hasBatchDim = rank == 3;
   int mIndex = 0 + hasBatchDim;
 
-  auto kWidth = dotMfmaLayout.getKWidth();
+  int32_t kWidth = dotMfmaLayout.getKWidth();
+  auto kDim = dotMfmaLayout.getOpIdx() == 0 ? rank - 1 : rank - 2;
+  int32_t kSize = shape[kDim];
   auto warpsPerCTA = mfmaLayout.getWarpsPerCTA();
 
-  if (kWidth != 8 || warpsPerCTA[mIndex] != 1) {
-    return std::nullopt;
-  }
-
   MLIRContext *ctx = dotMfmaLayout.getContext();
   SmallVector<StringAttr> outDimNames = standardOutDimNames(ctx, rank);
 
   StringAttr kRegister = S("register");
   StringAttr kLane = S("lane");
+  StringAttr kWarp = S("warp");
 
+  // register order
+  // operand A: [1, 0] / [2, 1, 0]
+  // operand B: [0, 1] / [1, 2, 0]
+  // for both cases it is [k, nonk]/[k, nonk, batch]
   SmallVector<unsigned> order = triton::gpu::getOrder(dotMfmaLayout);
-  auto tileLayout = LinearLayout::empty();
+  // warp order
+  // common for both operand A and B: [0, 1] / [0, 1, 2]
+  // in both cases it is [M dim, N dim]/[batch, M dim, N dim]
+  SmallVector<unsigned> warpOrder = triton::gpu::getWarpOrder(dotMfmaLayout);
```

Comment:
> I think you always want the batch dimension to be the slowest running

Right, this is how current implementation works.

> The outer strides property is not a huge deal for AMD for all I know, but it's consistent with nvidia's wgmma's instructions, which extend the warps along the M dimension and have different sizes in the N dimension.

I don't understand this part. I am not very familiar with WGMMA instructions, but it seems they work different from MFMA. MFMA is inter-warp instructions, i.e. there are no communication between warps.

This is why we need to preserve same order for both arguments. Consider following example with 4 warps and order [1, 0] (M is 0 dim, N is 1 dim):

![image](https://github.com/user-attachments/assets/fe75fed5-82d7-43e4-ad79-e72bfe898475)

Each warp holds part of result C tensor and requires to hold all elements of operands in corresponding rows and columns. i.e. to compute warp 1 values warp 1 should hold 0 row and 1 column.

If we provide different orders to operands, like this:
![image](https://github.com/user-attachments/assets/ba758aad-365c-4694-bf85-9ec511236321)
There are no way to compute part of the matrix above and below main diagonal.

> same as above

I would like to use general functions, like "getWarpOrder", because `getOrderForDotOperand` looks like an internal implementation function.

Reply:
ah, you are right, the current behaviour is correct. I had completely missed that.